File: leetcode/3_longest-substring-without-repeating-characters/set.go
func lengthOfLongestSubstring(s string) int {
	if len(s) <= 1 {
		return len(s)
	}
	ans := 0
	start := 0
	set := map[byte]struct{}{}
	for i, r := range []byte(s) {
		if _, exists := set[r]; exists {
			for s[start] != r {
				delete(set, s[start])
				start++
			}
			start++
		} else {
			set[r] = struct{}{}
			if ans < i-start+1 {
				ans = i - start + 1
			}
		}
	}
	return ans
}

File: leetcode/3_longest-substring-without-repeating-characters/set.cpp
// https://leetcode.com/problems/longest-substring-without-repeating-characters/submissions/
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int ans = 0;
        unordered_set<char> char_set;
        int left = 0;
        for (int right = 0; right < s.size(); right++) {
            if (!char_set.count(s[right])) {
                char_set.insert(s[right]);
                ans = max(ans, right - left + 1);
            } else {
                while (s[left] != s[right]) {
                    char_set.erase(s[left]);
                    left++;
                }
                left++;
            }
        }
        return ans;
    }
};
